src/devices/device_models/common/range_constraint.jl

```diff
@@ -214,3 +232,43 @@ function reserve_device_semicontinuousrange(canonical::CanonicalModel,
     return
 
  end
+
+@doc raw"""
+    device_range_expression(canonical::CanonicalModel,
+                        devices::Vector{T},
+                        exp_name::Symbol,
+                        var_name::Symbol) where {T<:PSY.Component}
+
+Constructs expression for min/max range constraint from device variable.
+
+# Expresion
+
+``` expression_container[device_name, time_index] =+ variable ```
+
+# Arguments
+* canonical::CanonicalModel : the canonical model built in PowerSimulations
+* devices::Vector{T} : contains devices
+* exp_name::Symbol : name of the expresion that makes up the LHS of a constraint
+* var_name::Symbol : the name of the continuous variable
+"""
+
+function device_range_expression!(canonical::CanonicalModel,
+                    devices::Vector{T},
+                    exp_name::Symbol,
+                    var_name::Symbol) where {T<:PSY.Component}
+
+    time_steps = model_time_steps(canonical)
+    var = PSI.var(canonical, var_name)
+    expression_cont = exp(canonical, exp_name)
+    for t in time_steps , d in devices
+        name = PSY.get_name(d)
+        if isassigned(expression_cont, name,t)
```

Comment:
I create the expression array for a constraint (say active range) for all Generators, and if a device is contributing to a service then it would require to add that service variable to the device range constraint. But the same device can be contributing to multiple services, so its possible that the array is already populated with variables from different services or is empty and this is the first variable that is being added to the expression.

Reply:
I didn't find a better way to do this

Reply:
We have a function for this.